feat(core): add type checking to @Pipe (#11126)

manumiranda opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
As stated by #11126, `@Pipe` decorator fails at execution time when it decorates a class that doesn't fulfill `PipeTransform` interface signature.

**What is the new behavior?**
The code will not compile and will show an error message.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

**Other information**:

Ensures at compile time that only classes that satisfy PipeTrasform interface
can be decorated by @Pipe.

closes angular/angular#11126

manumiranda committed:
feat(core): add type checking to @Pipe (#11126)

Ensures at compile time that only classes that satisfy PipeTrasform interface
can be decorated by @Pipe.

closes angular/angular#11126

robwormald commented on modules/@angular/compiler/test/metadata_resolver_spec.ts:
```diff
@@ -77,6 +77,7 @@ export function main() {
          inject([CompileMetadataResolver], (resolver: CompileMetadataResolver) => {
            @Pipe({name: 'somePipe'})
            class SomePipe {
+            transform(){};
```
this should test that it works with arguments, eg `transform(value:any)`, maybe add a couple that test various scenarios.

robwormald commented on modules/@angular/compiler/test/pipe_resolver_mock_spec.ts:
```diff
@@ -36,4 +36,5 @@ export function main() {
 
 @Pipe({name: 'somePipe'})
 class SomePipe {
+  transform(){};
```
again should have an argument

robwormald commented:
this looks great, thanks for the PR. added a couple of comments.

once you've updated, you'll want to run `gulp format:enforce` - it will print out a command for you to run. after that run `gulp public-api:update` to update the public API golden version.

manumiranda commented:
@robwormald thanks for the help! I'm just following your tips to get this corrected asap =)

manumiranda commented:
@robwormald I don't know if I must notify you or something when all suggested changes are applied or if I only have to wait...

vicb commented:
Could you please rebase and squash your commits ?

manumiranda commented:
I rebased and squashed my commits following the contributor's guide but now its failing on `npm install`... I don't really know what to do now as this PR worked and passed all tests previously. Should I update my branch at the current state of the project?

Any help would be greatly apreciated

laco0416 commented:
Isn't this a breaking change? I think it's possible to cause errors at compilation in existing projects.

chuckjaz commented:
@manumiranda We can now take this change. Can you rebase it on master?

manumiranda commented:
I've been taking a look to the new codebase and there are too many changes to for my patch to work as it is. I should take some time to try to fix it but time is a problem for me right now =(

DzmitryShylovich commented:
I would expect signature like this:

```
PipeTransform<T, R> {
    transform(value: T, args: any[]): R {}
}
```

vicb commented:
> I've been taking a look to the new codebase and there are too many changes to for my patch to work as it is. I should take some time to try to fix it but time is a problem for me right now =(

Closing for now then... I you find a solution to your problem, I would also be interested ;)
